controllers/pruner/pruner_controller.go

```diff
@@ -0,0 +1,338 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	dwv2 "github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+	controllerv1alpha1 "github.com/devfile/devworkspace-operator/apis/controller/v1alpha1"
+	"github.com/devfile/devworkspace-operator/pkg/conditions"
+	"github.com/devfile/devworkspace-operator/pkg/config"
+
+	"github.com/operator-framework/operator-lib/prune"
+	"github.com/robfig/cron/v3"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// DevWorkspacePrunerReconciler reconciles DevWorkspace objects for pruning purposes.
+type DevWorkspacePrunerReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+
+	cron *cron.Cron
+}
+
+func shouldReconcileOnUpdate(e event.UpdateEvent, log logr.Logger) bool {
+	log.Info("DevWorkspaceOperatorConfig update event received")
+	oldConfig, ok := e.ObjectOld.(*controllerv1alpha1.DevWorkspaceOperatorConfig)
+	if !ok {
+		return false
+	}
+	newConfig, ok := e.ObjectNew.(*controllerv1alpha1.DevWorkspaceOperatorConfig)
+	if !ok {
+		return false
+	}
+
+	oldCleanup := oldConfig.Config.Workspace.CleanupCronJob
+	newCleanup := newConfig.Config.Workspace.CleanupCronJob
+
+	differentBool := func(a, b *bool) bool {
+		switch {
+		case a == nil && b == nil:
+			return false
+		case a == nil || b == nil:
+			return true
+		default:
+			return *a != *b
+		}
+	}
+	differentInt32 := func(a, b *int32) bool {
+		switch {
+		case a == nil && b == nil:
+			return false
+		case a == nil || b == nil:
+			return true
+		default:
+			return *a != *b
+		}
+	}
+
+	if oldCleanup == nil && newCleanup == nil {
+		return false
+	}
+	if (oldCleanup == nil && newCleanup != nil) || (oldCleanup != nil && newCleanup == nil) {
+		return true
+	}
+	if differentBool(oldCleanup.Enable, newCleanup.Enable) {
+		return true
+	}
+	if differentBool(oldCleanup.DryRun, newCleanup.DryRun) {
+		return true
+	}
+	if differentInt32(oldCleanup.RetainTime, newCleanup.RetainTime) {
+		return true
+	}
+	return oldCleanup.Schedule != newCleanup.Schedule
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *DevWorkspacePrunerReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	log := r.Log.WithName("setupWithManager")
+	log.Info("Setting up DevWorkspacePrunerReconciler")
+
+	configPredicate := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return shouldReconcileOnUpdate(e, log)
+		},
+		CreateFunc:  func(e event.CreateEvent) bool { return true },
+		DeleteFunc:  func(e event.DeleteEvent) bool { return false },
```

Comment:
If the cronjob is enabled with a DWOC, and then if we delete the DWOC, will the cronjob still be enabled?